Feature--Backend--Request-Controller

ChadNaps opened the pull request:
### What this PR does (required):
- Adds controller for requests according to the following endpoints
    - GET /request/load - Gets list of requests for currently logged-in user
    - POST /request/new - Creates a new request and saves it in the database
    - PATCH /request/update - Updates provided boolean values for accepted, declined, and paid fields of Request object currently saved in database
- Adds helper functions in utils/ for potential use throughout the project

### Any information needed to test this feature (required):
- Add dummy user to database by running test "POST /request/new"
- Check for dummy user by running test "GET /request/load"

### Any issues with the current functionality (optional):
- The testing is incomplete

ChadNaps committed:
Add request controller boilerplate

ChadNaps committed:
Add request controller routes

ChadNaps committed:
Update .gitignore

ChadNaps committed:
Update /request/new route

Uses POST instead of GET.

ChadNaps committed:
Add helperFunctions.js

Contains functions that may prove useful in other parts of the app.

ChadNaps committed:
Add incomplete first pass of /request/new route

Still needs to be updated once PR #53 is accepted.

ChadNaps committed:
Update controllers/request.js

Small bugfix - res not visible in validateQuery().
Saves to open database.

ChadNaps committed:
Update utils/helperFunctions.js

- Add extra validation.
- Minor bug fix - Errors in isArrayBoolean() improperly thrown.
- Add explanation for isObjectIdValid

ChadNaps committed:
Update controllers/request.js

Add ability for logged-in user to retrieve all their requests.

ChadNaps committed:
Merge branch 'main' into Feature--Backend--Request-Controller

# Conflicts:
#	.gitignore

ChadNaps committed:
Update controllers/request.js

Add address model to /request/new route.
Small refactor of helperFunctions.js

ChadNaps committed:
Update routes/request.js

Changes /request/update route to PATCH instead of GET.

ChadNaps committed:
Add more helperFunctions to utils/helperFunctions.js

These functions are intended to be used in controller/request.js logic.

ChadNaps committed:
Add /request/update logic to controllers/request.js

ChadNaps committed:
Refactor controllers/request.js and utils/helperFunctions.js

ChadNaps committed:
Fix server/test/request_model.test.js

Tests the final version of the request model properly.

ChadNaps committed:
Add initial code for request_controller.test.js

ChadNaps committed:
Bug fix

Route requests/new now accepts data from both a query and body.

ChadNaps committed:
Update request_controller.test.js

Intermediary step to fix test.

moffatethan commented on server/controllers/auth.js:
```diff
@@ -2,8 +2,8 @@ const User = require("../models/User");
 const generateToken = require("../utils/generateToken");
-const dotenv = require('dotenv').config({path:__dirname+'/./../.env'})
-const loginHelper = require('../helpers/loginHelper');
+const dotenv = require("dotenv").config({ path: __dirname + "/./../.env" });
```
We wouldn't need to use this here, once the env is attached to the process from the `bin/www` startup script it will be available anywhere else in the codebase. 

If you could just remove it - I know you weren't the one who added it but that would be appreciated 💯

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+  validateQueryForNewRequest,
+} = require("../utils/helperFunctions");
+
+// @route GET /request/load
```
Not a fan of this route name `/requests` is fine

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+// @desc Get list of requests for logged-in user
+// @access Private
+exports.loadRequests = asyncHandler(async (req, res, next) => {
+  const uid = req.user.id;
+  let requests;
+
+  if (uid) {
+    requests = await Request.find({
+      userId: uid,
+    });
+  }
```
So logically, this route will have the protect middleware. In that case, `req.user.id` will never be undefined and if the request doesn't have the auth cookie it will never hit this controller so we can clean this up a lot. We can just query for the users requests and store that in a constant then return a successful result since even empty arrays denote just no requests.

Hope this makes sense 👍🏻

ChadNaps replied:
Sure, I see what you're saying. Makes sense! I actually grappled with that decision at first too, but decided to add the extra check in case, for example, the authentication were to go through a change, unwittingly exposing the route. Ultimately not a big deal either way on this one I think, but I wanted to explain the thought process here anyway.

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+  res.status(200).json(requests);
+});
+
+// @route POST /request/new
```
Just `POST /requests`

moffatethan replied:
It's generally not common to explicitly define `/new` in REST APIs, reason is the HTTP method tells what is going to happen. 🙂

ChadNaps replied:
Gotcha, that makes sense. I've never really known what the preffered convention is since I've only ever really written routes for myself. Thanks for clearing it up for me!

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+
+// @route POST /request/new
+// @desc Create a new request
+// @access Public
```
Access needs to be private instead, we only want authenticated users to make requests.

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+  validateQueryForNewRequest(res, data);
+
+  const request = new Request({
+    userId: data.userId,
```
Note for future: We'll need to create a new PaymentSchema instance when a request is created, a lot of the request properties may be transferred to the payment.

ChadNaps replied:
Good note, thanks!

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+exports.newRequest = asyncHandler(async (req, res, next) => {
+  const { validateQueryForNewRequest } = require("../utils/helperFunctions");
+
+  let data;
+  if (Object.keys(req.query).length > 0) {
+    data = req.query;
+  } else {
+    data = req.body;
+  }
```
Could you elaborate on this conditional's purpose?

moffatethan replied:
Is it for allowing a new request to be created via query parameters? I am not a huge fan of this, we should enforce the request payload be sent down in the body.

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+    },
+  });
+
+  if (!(await request.save())) {
```
Not a fan of this, could we just do a try/catch?

moffatethan replied:
Or if you don't want to specifically handle failed save request - we could just let our async handler catch the error

ChadNaps replied:
True! I'll just let the default error handling catch it.

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+  });
+
+  if (!(await request.save())) {
+    res.status(500).send("There was an error saving your request");
```
Also, could we do JSON responses here?

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+    res.status(500).send("There was an error saving your request");
+  }
+
+  res.status(200).send("Successfully saved request!");
```
Not sure on the response - wouldn't it be better to send back the newly created request resource for the client side to use?

ChadNaps replied:
Ha ha, I was really not feeling awesome about these responses either. I'm glad you offered better solutions!

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+  res.status(200).send("Successfully saved request!");
+});
+
+// @route PATCH /request/update
```
Same situation as the others - `PATCH /requests/:requestId` will be more RESTful

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+// @desc Update request with approved or declined
+// @access Private
+exports.updateRequest = asyncHandler(async (req, res, next) => {
+  const requestId = req.query.requestId;
```
Let's pass by URL params and not query parameters. Not really something that is supposed to be passed via query params.

moffatethan commented on server/controllers/request.js:
```diff
@@ -0,0 +1,95 @@
+
+  request = updateRequest(request, fieldsToChange, req);
+
+  if (!(await request.save())) {
+    res.status(500).send("There was an error saving your updated request");
+  }
```
Same as before - not a big fan of this

moffatethan commented on server/routes/request.js:
```diff
@@ -0,0 +1,16 @@
+  updateRequest,
+} = require("../controllers/request");
+
+router.route("/load").get(protect, loadRequests);
+
+router.route("/new").post(newRequest);
+
+router.route("/update").patch(protect, updateRequest);
```
Let's update the names of this

moffatethan commented on server/utils/helperFunctions.js:
```diff
@@ -0,0 +1,146 @@
+/*
+ * Contents
+ * Public:
+ *  - checkQueryForEmptyFields()
+ *  - updateRequest()
+ *  - validateQuery()
+ * Private:
+ *  - updateRequestField()
+ *  - isObjectIdValid()
+ *  - isArrayBoolean()
+ */
```
Nice - but I don't think we need this here since you're already defining documentation in the helpers.

moffatethan commented on server/utils/helperFunctions.js:
```diff
@@ -0,0 +1,146 @@
+/*
```
I like that you considered the validation on the incoming inputs - it's excellent to see. However, I don't want to have to maintain helpers like this.

Rather, I think it would make much more sense to utilize a library like https://express-validator.github.io/docs/ which is already being used in the server so I just want to keep it consistent.

Code example: https://github.com/hatchways/team-oak/blob/main/server/validate.js

ChadNaps replied:
Got this done. I'm not going to push the changes until I have a chance to double check and make sure migrating to middlewear works as intended.

moffatethan reviewed (changes requested):
Hey Chad - decent work here, I like the documentation that you've been writing on each wrote. Just some comments on clean up and logic concerns.

Please take a look at all the comments and spend some time on this a bit more. Overall, you're on the right track. 

Thanks! 🙂

ChadNaps committed:
Correct changes requested from PR #74

Remove unnecessary helper function

Update route PATCH /requests/:requestId

Fixes function from previous commit.

Update input validation for POST /requests

Moves validation from local function to express middleware, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789920389

Refactor utils/helperFunctions.js

Remove unnecessary comment, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789918982

Update PATCH /requests/:requestId route

Removes unnecessary error checking, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789918445

Update PATCH /requests/:requestId route

Changes `req.query.requestId` to `req.params.requestId`, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789918267

Update PATCH /request/update route

Changes PATCH /request/update to PATCH /requests/:requestId, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789917934

Update POST /requests route

Route returns server-created request object to client, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789916900

Update POST /requests route

Remove unnecessary error check, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789916175

Update POST /requests route

Remove support for request creation via `req.query` parameters and explicitly enforce request be in `req.body`, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789913947

Update POST /requests route

Changed from Public to Private access, as per: https://github.com/hatchways/team-oak/pull/74#discussion_r789912490

Update POST /requests route

Change from POST /request/new to POST /requests. Done as per request: https://github.com/hatchways/team-oak/pull/74#discussion_r789911753

Update controllers/request.js

Removed extra auth check for GET /requests route, as per request: https://github.com/hatchways/team-oak/pull/74#discussion_r789911574

Update GET /requests route

Change from /request/load to /requests. Done as per request: https://github.com/hatchways/team-oak/pull/74#discussion_r789909714

Refactor controllers/auth.js

Remove dotenv.

ChadNaps committed:
Update POST /requests and PATCH /requests/:requestId

Change from incorrectly using `req.params` and `req.query` fields to using `req.body`. Also fixes broken validation for POST /requests route.

ChadNaps committed:
Merge branch 'main' into Feature--Backend--Request-Controller

moffatethan reviewed (approved):
Good work!
